controllers/nodemodulesconfig_reconciler.go

```diff
@@ -0,0 +1,681 @@
+
+	logger.Info("Syncing status")
+
+	pods, err := w.pm.ListWorkerPodsOnNode(ctx, nmcObj.Name)
+	if err != nil {
+		return fmt.Errorf("could not list worker pods for NodeModulesConfig %s: %v", nmcObj.Name, err)
+	}
+
+	logger.V(1).Info("List worker Pods", "count", len(pods))
+
+	if len(pods) == 0 {
+		return nil
+	}
+
+	patchFrom := client.MergeFrom(nmcObj.DeepCopy())
+	errs := make([]error, 0, len(pods))
+
+	for _, p := range pods {
+		podNSN := types.NamespacedName{Namespace: p.Namespace, Name: p.Name}
+
+		modNamespace := p.Namespace
+		modName := p.Labels[constants.ModuleNameLabel]
+		phase := p.Status.Phase
+
+		logger := logger.WithValues("pod name", p.Name, "pod phase", p.Status.Phase)
+
+		logger.Info("Processing worker Pod")
+
+		status := kmmv1beta1.NodeModuleStatus{
+			Namespace: modNamespace,
+			Name:      modName,
+		}
+
+		deletePod := false
+		statusDeleted := false
+
+		switch phase {
+		case v1.PodFailed:
+			deletePod = true
+		case v1.PodSucceeded:
+			deletePod = true
+
+			if p.Labels[actionLabelKey] == WorkerActionUnload {
+				nmc.RemoveModuleStatus(&nmcObj.Status.Modules, modNamespace, modName)
+				deletePod = true
+				statusDeleted = true
+				break
+			}
+
+			config := kmmv1beta1.ModuleConfig{}
+
+			if err = yaml.UnmarshalStrict([]byte(p.Annotations[configAnnotationKey]), &config); err != nil {
+				errs = append(
+					errs,
+					fmt.Errorf("%s: could not unmarshal the ModuleConfig from YAML: %v", podNSN, err),
+				)
+
+				continue
+			}
+
+			status.Config = &config
+
+			podLTT := GetContainerStatus(p.Status.ContainerStatuses, workerContainerName).
+				State.
+				Terminated.
+				FinishedAt
+
+			status.LastTransitionTime = &podLTT
+
+			deletePod = true
+		case v1.PodPending, v1.PodRunning:
+			status.InProgress = true
+			// TODO: if the NMC's spec changed compared to the Pod's config, recreate the Pod
+		default:
+			errs = append(
+				errs,
+				fmt.Errorf("%s: unhandled Pod phase %q", podNSN, phase),
+			)
+		}
+
+		if deletePod {
+			if err = w.pm.DeletePod(ctx, &p); err != nil {
+				errs = append(
+					errs,
+					fmt.Errorf("could not delete worker Pod %s: %v", podNSN, errs),
+				)
+
+				continue
+			}
+		}
+
+		if !statusDeleted {
+			nmc.SetModuleStatus(&nmcObj.Status.Modules, status)
+		}
+	}
+
+	if err = errors.Join(errs...); err != nil {
+		return fmt.Errorf("encountered errors while reconciling NMC %s status: %v", nmcObj.Name, err)
+	}
+
+	if err = w.client.Status().Patch(ctx, nmcObj, patchFrom); err != nil {
+		return fmt.Errorf("could not patch NodeModulesConfig %s status: %v", nmcObj.Name, err)
+	}
+
+	return nil
+}
+
+const (
+	configFileName = "config.yaml"
+	configFullPath = volMountPoingConfig + "/" + configFileName
+
+	volNameConfig       = "config"
+	volMountPoingConfig = "/etc/kmm-worker"
+)
+
+//go:generate mockgen -source=nodemodulesconfig_reconciler.go -package=controllers -destination=mock_nodemodulesconfig_reconciler.go PodManager
+
+type PodManager interface {
```

Comment:
WDYT about moving this to a dedicated package?

Reply:
What for? It is only used in the reconciler.

Reply:
since we are moving build/sign to use pods also, won't it be more convenient to have a dedicated package?

Reply:
What commonality is there between build and worker Pods?

Reply:
that it is dealing with pods?...  why is pod interface needs to be at this file? It does not deal with reconciliation logic, but with creating Pod objects based on data

Reply:
I would only extract common code. This code is not common; it is only used once, in the reconciler.
If we extract everything, we are going to end up with a lot of packages in `internal`.

Reply:
i am ok with moving that decision to 2.1